libs/vault/src/services/at-risk-password-callout.service.ts

```diff
@@ -53,11 +53,66 @@ export class AtRiskPasswordCalloutService {
     );
   }
 
+  shouldShowCompletionBanner$(userId: UserId): Observable<boolean> {
+    return combineLatest([
+      this.pendingTasks$(userId),
+      this.atRiskPasswordState(userId).state$,
+    ]).pipe(
+      map(([pendingTasks, state]) => {
+        const hasPendingTasks = pendingTasks.length > 0;
+        const {
+          showTasksCompleteBanner = false,
+          hadPendingTasks = false,
+          tasksBannerDismissed = false,
+        } = state ?? {};
+
+        // Show banner if
+        // user has resolved all pending tasks
+        // user had showTasksCompleteBanner set to true and hasn't received new tasks
+        // user had showTasksCompleteBanner set to true and hasn't dismissed
+        return (
+          !hasPendingTasks && (showTasksCompleteBanner || hadPendingTasks) && !tasksBannerDismissed
+        );
+      }),
+    );
+  }
+
   atRiskPasswordState(userId: UserId): SingleUserState<AtRiskPasswordCalloutData> {
     return this.stateProvider.getUser(userId, AT_RISK_PASSWORD_CALLOUT_KEY);
   }
 
   updateAtRiskPasswordState(userId: UserId, updatedState: AtRiskPasswordCalloutData): void {
     void this.atRiskPasswordState(userId).update(() => updatedState);
   }
+
+  updatePendingTasksState(userId: UserId, currentTaskCount: number): void {
+    this.atRiskPasswordState(userId)
+      .state$.pipe(
+        take(1),
+        tap((currentState) => {
+          let updateObject: AtRiskPasswordCalloutData | null = null;
+
+          // If user had pending tasks and resolved all, show banner
+          if (currentState?.hadPendingTasks && currentTaskCount === 0) {
+            updateObject = {
+              hadPendingTasks: false,
+              showTasksCompleteBanner: true,
+              tasksBannerDismissed: false,
+            };
+            // If user has pending tasks, set hadPendingTasks to true
+          } else if (currentTaskCount > 0) {
+            updateObject = {
+              hadPendingTasks: true,
+              showTasksCompleteBanner: false,
+              tasksBannerDismissed: false,
+            };
+          }
+
+          if (updateObject) {
+            this.updateAtRiskPasswordState(userId, updateObject);
+          }
+        }),
+      )
+      .subscribe();
+  }
```

Comment:
suggestion: just use the update function directly or you'll risk introducing race-conditions depending on how this function decides to handle multiple calls and creations/subscriptions to its internal observable

```suggestion
  updatePendingTasksState(userId: UserId, currentTaskCount: number): void {
    this.atRiskPasswordState(userId).update((currentState) => {
      let updateObject;

      // If user had pending tasks and resolved all, show banner
      if (currentState?.hadPendingTasks && currentTaskCount === 0) {
        updateObject = {
          hadPendingTasks: false,
          showTasksCompleteBanner: true,
          tasksBannerDismissed: false,
        };
        // If user has pending tasks, set hadPendingTasks to true
      } else if (currentTaskCount > 0) {
        updateObject = {
          hadPendingTasks: true,
          showTasksCompleteBanner: false,
          tasksBannerDismissed: false,
        };
      } else {
        updateObject = currentState;
      }

      return updateObject;
    });
  }
```